File: mbw/src/main/java/com/mycelium/wallet/activity/view/ValueKeyboard.java
package com.mycelium.wallet.activity.view;


import android.content.Context;
import android.support.v7.widget.GridLayout;
import android.util.AttributeSet;
import android.util.TypedValue;
import android.view.View;
import android.widget.TextView;

import com.mycelium.wallet.R;
import com.mycelium.wallet.Utils;

import java.math.BigDecimal;

public class ValueKeyboard extends GridLayout {
    public final static int DEL = -1;
    public final static int DOT = -2;
    public final static int MAX_DIGITS_BEFORE_DOT = 9;

    InputListener inputListener = null;
    TextView inputTextView = null;

    int maxDecimals = 0;

    public ValueKeyboard(Context context) {
        super(context);
    }

    public ValueKeyboard(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    public ValueKeyboard(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }

    public TextView getInputTextView() {
        return inputTextView;
    }

    public void setMaxDecimals(int maxDecimals) {
        this.maxDecimals = maxDecimals;
        value.setEntry(value.getEntryAsBigDecimal(), maxDecimals);
        updateDotBtn();
    }

    BigDecimal spendableValue = BigDecimal.ZERO;

    public void setSpendableValue(BigDecimal spendableValue) {
        this.spendableValue = spendableValue;
        updateMaxBtn();
    }

    BigDecimal maxValue;

    public void setMaxValue(BigDecimal maxValue) {
        this.maxValue = maxValue;
    }

    public void setInputTextView(TextView inputTextView) {
        this.inputTextView = inputTextView;
    }

    NumberEntry value = new NumberEntry(maxDecimals, "", new EntryChange() {
        @Override
        public void entryChange(String entry, boolean wasSet) {
            if (inputTextView != null) {
                inputTextView.setText(entry);
            }
        }
    });

    public void setEntry(String val) {
        value = new NumberEntry(maxDecimals, val, new EntryChange() {
            @Override
            public void entryChange(String entry, boolean wasSet) {
                if (inputTextView != null) {
                    inputTextView.setText(entry);
                }
            }
        });
    }

    public void setInputListener(InputListener inputListener) {
        this.inputListener = inputListener;
    }

    @Override
    protected void onFinishInflate() {
        super.onFinishInflate();
        for (int i = 0; i < getChildCount(); i++) {
            View view = getChildAt(i);
            view.setOnClickListener(new OnClickListener() {
                @Override
                public void onClick(View view) {
                    if (view.getId() == R.id.btn_max) {
                        value.setEntry(spendableValue, maxDecimals);
                    } else if (view.getId() == R.id.btn_backspace) {
                        value.clicked(DEL);
                    } else if (view.getId() == R.id.btn_dot) {
                        value.clicked(DOT);
                    } else if (view.getId() == R.id.btn_done) {
                        done();
                    } else if (view.getId() == R.id.btn_copy) {
                        String clipboardString = Utils.getClipboardString(getContext());
                        if (!clipboardString.isEmpty()) {
                            try {
                                value.setEntry(new BigDecimal(clipboardString), maxDecimals);
                            } catch (NumberFormatException ignore) {
                            }
                        }
                    } else if (view instanceof TextView) {
                        value.clicked(Integer.parseInt(((TextView) view).getText().toString()));
                    }
                }
            });
        }

        updateDotBtn();
        updateMaxBtn();

        findViewById(R.id.btn_backspace).setOnLongClickListener(new OnLongClickListener() {
            @Override
            public boolean onLongClick(View view) {
                value.clear();
                return true;
            }
        });
    }

    public void setMaxText(String text, float size) {
        TextView textView = findViewById(R.id.btn_max);
        textView.setText(text);
        textView.setTextSize(TypedValue.COMPLEX_UNIT_SP, size);
    }

    public void setPasteVisibility(int visibility) {
        findViewById(R.id.btn_copy).setVisibility(visibility);
    }

    public void done() {
        setVisibility(View.GONE);
        if (inputListener != null) {
            inputListener.done();
        }
    }

    private void updateDotBtn() {
        findViewById(R.id.btn_dot).setEnabled(maxDecimals > 0);
    }

    private void updateMaxBtn() {
        findViewById(R.id.btn_max).setVisibility(spendableValue.equals(BigDecimal.ZERO) ? View.INVISIBLE : View.VISIBLE);
    }

    public interface InputListener {
        void input(CharSequence sequence);

        void max();

        void backspace();

        void done();
    }

    public static class SimpleInputListener implements InputListener {

        @Override
        public void input(CharSequence sequence) {

        }

        @Override
        public void max() {

        }

        @Override
        public void backspace() {

        }

        @Override
        public void done() {

        }
    }

    interface EntryChange {
        void entryChange(String entry, boolean wasSet);
    }

    class NumberEntry {
        int _maxDecimals;
        String entry;
        EntryChange entryChange;

        public NumberEntry(int _maxDecimals, String entry, EntryChange entryChange) {
            this._maxDecimals = _maxDecimals;
            this.entry = entry;
            this.entryChange = entryChange;
            if (!entry.isEmpty()) {
                try {
                    entry = new BigDecimal(entry).toPlainString();
                } catch (Exception e) {
                    entry = "";
                }
            }

//            if (_maxDecimals > 0) {
//                setClickListener(_llNumberEntry.findViewById(R.id.btDot) as Button, DOT)
//            } else {
//                (_llNumberEntry.findViewById(R.id.btDot) as Button).text = ""
//            }
//            setClickListener(_llNumberEntry.findViewById(R.id.btZero) as Button, 0)
//            setClickListener(_llNumberEntry.findViewById(R.id.btDel) as Button, DEL)

//            _llNumberEntry.findViewById(R.id.btDel).setOnLongClickListener {
//                entry = ""
//                _listener.onEntryChanged(entry, false)
//                true
//            }
        }

        void clear() {
            entry = "";
            entryChange.entryChange(entry, false);
        }

        void clicked(int digit) {
            if (entry.equals("0")) {
                entry = "";
            }
            if (digit == DEL) {
                // Delete Digit
                if (!entry.isEmpty()) {
                    entry = entry.substring(0, entry.length() - 1);
                }
            } else if (digit == DOT) {
                // Do we already have a dot?
                if (hasDot()) {
                    return;
                }
                if (_maxDecimals == 0) {
                    return;
                }
                if (entry.isEmpty()) {
                    entry = "0.";
                } else {
                    entry += '.';
                }
            } else {
                // Append Digit
                if (digit == 0 && "0".equals(entry)) {
                    // Only one leading zero
                    return;
                }
                if (hasDot()) {
                    if (decimalsAfterDot() >= _maxDecimals) {
                        // too many decimals
                        return;
                    }
                } else {
                    if (decimalsBeforeDot() >= MAX_DIGITS_BEFORE_DOT) {
                        return;
                    }
                }
                if (maxValue == null || new BigDecimal(entry + digit).compareTo(maxValue) <= 0) {
                    entry = entry + digit;
                }
            }
            entryChange.entryChange(entry, false);
        }

        private boolean hasDot() {
            return entry.indexOf('.') != -1;
        }

        private int decimalsAfterDot() {
            int dotIndex = entry.indexOf('.');
            if (dotIndex == -1) {
                return 0;
            }
            return entry.length() - dotIndex - 1;
        }

        private int decimalsBeforeDot() {
            int dotIndex = entry.indexOf('.');
            if (dotIndex == -1) {
                return entry.length();
            }
            return dotIndex;
        }

        void setEntry(BigDecimal number, int maxDecimals) {
            _maxDecimals = maxDecimals;
            if (number == null || number.compareTo(BigDecimal.ZERO) == 0) {
                entry = "";
            } else {
                entry = number.setScale(_maxDecimals, BigDecimal.ROUND_HALF_DOWN).stripTrailingZeros().toPlainString();
            }
            entryChange.entryChange(entry, true);
        }


        public BigDecimal getEntryAsBigDecimal() {
            if (entry.isEmpty()) {
                return BigDecimal.ZERO;
            }
            if ("0.".equals(entry)) {
                return BigDecimal.ZERO;
            }
            try {
                return new BigDecimal(entry);
            } catch (NumberFormatException e) {
                return BigDecimal.ZERO;
            }
        }
    }
}

File: mbw/src/main/java/com/mycelium/wallet/activity/view/ValueKeyboard.kt
package com.mycelium.wallet.activity.view

import android.content.Context
import android.util.AttributeSet
import android.util.TypedValue
import android.view.View
import android.widget.TextView
import androidx.constraintlayout.widget.ConstraintLayout
import com.mycelium.wallet.R
import com.mycelium.wallet.Utils
import java.math.BigDecimal

class ValueKeyboard : ConstraintLayout {
    constructor(context: Context) : super(context)
    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs)
    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) : super(context, attrs, defStyleAttr)

    var inputListener: InputListener? = null
    var errorListener: ErrorListener? = null

    var inputTextView: TextView? = null
        set(v) {
//            field?.setTextColor(resources.getColor(R.color.white))
            field = v
        }

    var maxDecimals = 0
        set(v) {
            field = v
            value.setEntry(value.entryAsBigDecimal, v)
            updateDotBtn()
        }

    var spendableValue: BigDecimal? = null
        set(v) {
            field = v
            updateMaxBtn()
        }

    var maxValue: BigDecimal? = null
        set(v) {
            field = v
            checkErrors()
        }

    var minValue: BigDecimal? = null
        set(v) {
            field = v
            checkErrors()
        }

    var value = NumberEntry(maxDecimals, "", object : EntryChange {
        override fun entryChange(entry: String, wasSet: Boolean) {
            setToTextView(entry)
        }
    })

    fun setEntry(entry: String) {
        value = NumberEntry(maxDecimals, entry, object : EntryChange {
            override fun entryChange(entry: String, wasSet: Boolean) {
                setToTextView(entry)
            }
        })
    }

    private fun setToTextView(entry: String) {
        inputTextView?.let { textView ->
            textView.text = entry
            checkErrors()
        }
    }

    private fun checkErrors() = try {
        val entryValue = value.entryAsBigDecimal
        if (maxValue != null && maxValue!! < entryValue && entryValue != BigDecimal.ZERO) {
//            inputTextView?.setTextColor(resources.getColor(R.color.sender_recyclerview_background_red))
            errorListener?.maxError(maxValue!!)
        } else if (minValue != null && minValue!! > entryValue && entryValue != BigDecimal.ZERO) {
//            inputTextView?.setTextColor(resources.getColor(R.color.sender_recyclerview_background_red))
            errorListener?.minError(minValue!!)
        } else {
//            inputTextView?.setTextColor(resources.getColor(R.color.white))
            errorListener?.noError()
        }
    } catch (e: NumberFormatException) {
        errorListener?.formatError()
    }


    override fun onFinishInflate() {
        super.onFinishInflate()
        for (i in 0 until childCount) {
            val view = getChildAt(i)
            view.setOnClickListener { view ->
                when (view.id) {
                    R.id.btn_max -> {
                        value.setEntry(spendableValue, maxDecimals, true)
                    }
                    R.id.btn_backspace -> {
                        value.clicked(DEL)
                    }
                    R.id.btn_dot -> {
                        value.clicked(DOT)
                    }
                    R.id.btn_done -> {
                        done()
                    }
                    R.id.btn_copy -> {
                        val clipboardString = Utils.getClipboardString(context)
                        if (clipboardString.isNotEmpty()) {
                            try {
                                value.setEntry(BigDecimal(clipboardString), maxDecimals)
                            } catch (ignore: NumberFormatException) {
                            }
                        }
                    }
                    else -> {
                        if (view is TextView) {
                            value.clicked(view.text.toString().toInt())
                        }
                    }
                }
            }
        }
        updateDotBtn()
        updateMaxBtn()
        findViewById<View>(R.id.btn_backspace).setOnLongClickListener {
            value.clear()
            true
        }
    }

    fun setMaxText(text: String?, size: Float) {
        val textView = findViewById<TextView>(R.id.btn_max)
        textView.text = text
        textView.setTextSize(TypedValue.COMPLEX_UNIT_SP, size)
    }

    fun setPasteVisibility(visible: Boolean) {
        findViewById<View>(R.id.btn_copy).visibility = if (visible) View.VISIBLE else View.INVISIBLE
    }

    fun done() {
//        inputTextView?.setTextColor(resources.getColor(R.color.white))
        visibility = GONE
        inputListener?.done()
    }

    private fun updateDotBtn() {
        findViewById<View>(R.id.btn_dot).isEnabled = maxDecimals > 0
    }

    private fun updateMaxBtn() {
        findViewById<View>(R.id.btn_max).visibility = if (spendableValue == null) INVISIBLE else VISIBLE
    }

    interface InputListener {
        fun input(sequence: CharSequence?)
        fun max()
        fun backspace()
        fun done()
    }

    interface ErrorListener {
        fun maxError(maxValue: BigDecimal)
        fun minError(minValue: BigDecimal)
        fun formatError()
        fun noError()
    }


    open class SimpleInputListener : InputListener {
        override fun input(sequence: CharSequence?) {}
        override fun max() {}
        override fun backspace() {}
        override fun done() {}
    }

    interface EntryChange {
        fun entryChange(entry: String, wasSet: Boolean)
    }

    class NumberEntry(var _maxDecimals: Int, var entry: String, val entryChange: EntryChange) {

        init {
            if (entry.isNotEmpty()) {
                entry = try {
                    BigDecimal(entry).toPlainString()
                } catch (e: Exception) {
                    ""
                }
            }

//            if (_maxDecimals > 0) {
//                setClickListener(_llNumberEntry.findViewById(R.id.btDot) as Button, DOT)
//            } else {
//                (_llNumberEntry.findViewById(R.id.btDot) as Button).text = ""
//            }
//            setClickListener(_llNumberEntry.findViewById(R.id.btZero) as Button, 0)
//            setClickListener(_llNumberEntry.findViewById(R.id.btDel) as Button, DEL)

//            _llNumberEntry.findViewById(R.id.btDel).setOnLongClickListener {
//                entry = ""
//                _listener.onEntryChanged(entry, false)
//                true
//            }
        }

        fun clear() {
            entry = ""
            entryChange.entryChange(entry, false)
        }

        fun clicked(digit: Int) {
            if (entry == "0") {
                entry = ""
            }
            if (digit == DEL) {
                // Delete Digit
                if (entry.isNotEmpty()) {
                    entry = entry.substring(0, entry.length - 1)
                }
            } else if (digit == DOT) {
                // Do we already have a dot?
                if (hasDot()) {
                    return
                }
                if (_maxDecimals == 0) {
                    return
                }
                if (entry.isEmpty()) {
                    entry = "0."
                } else {
                    entry += '.'
                }
            } else {
                // Append Digit
                if (digit == 0 && "0" == entry) {
                    // Only one leading zero
                    return
                }
                if (hasDot()) {
                    if (decimalsAfterDot() >= _maxDecimals) {
                        // too many decimals
                        return
                    }
                } else {
                    if (decimalsBeforeDot() >= MAX_DIGITS_BEFORE_DOT) {
                        return
                    }
                }
                //                if (maxValue == null || new BigDecimal(entry + digit).compareTo(maxValue) <= 0) {
                entry = entry + digit
                //                }
            }
            entryChange.entryChange(entry, false)
        }

        private fun hasDot(): Boolean {
            return entry.indexOf('.') != -1
        }

        private fun decimalsAfterDot(): Int {
            val dotIndex = entry.indexOf('.')
            return if (dotIndex == -1) {
                0
            } else {
                entry.length - dotIndex - 1
            }
        }

        private fun decimalsBeforeDot(): Int {
            val dotIndex = entry.indexOf('.')
            return if (dotIndex == -1) {
                entry.length
            } else {
                dotIndex
            }
        }

        fun setEntry(number: BigDecimal?, maxDecimals: Int, zeroAcceptable:Boolean = false) {
            _maxDecimals = maxDecimals
            entry = if (number == null || (!zeroAcceptable && number.compareTo(BigDecimal.ZERO) == 0)) {
                ""
            } else {
                number.setScale(_maxDecimals, BigDecimal.ROUND_DOWN).stripTrailingZeros().toPlainString()
            }
            entryChange.entryChange(entry, true)
        }

        val entryAsBigDecimal: BigDecimal
            get() =
                when {
                    entry.isEmpty() -> BigDecimal.ZERO
                    "0." == entry -> BigDecimal.ZERO
                    else -> try {
                        entry.toBigDecimal()
                    } catch (e: NumberFormatException) {
                        BigDecimal.ZERO
                    }
                }
    }

    companion object {
        const val DEL = -1
        const val DOT = -2
        const val MAX_DIGITS_BEFORE_DOT = 9
    }
}